### DISCUSSION TITLE
file_help db entry support

### DISCUSSION THREAD

**Discussion by davewiththenicehat** [evennia/evennia#2426] [Evennia Engine Development]
@Griatch 
Within my project I am adding support to create database entries from file help entries.
After putting the option to do this in a variable in the setting file, is that something that would have a value to people other than myself?

I created a help refresh script that is called once a day. The functions within are also called in at_server_start (likely I would add the calls to what calls at_server_start, to leave at_server_start clean)
Help refresh replaces database entries if they do not match the file entries. It also removes database entries that no longer exist in the file help entries.
I'll be certain to remove the script within tearDown in unit tests. Otherwise it should produce unclean reactor assertions?

The usefulness of the entries being in the web UI meant I had to add the feature.

Thanks for all your time on the project.

**Comment by Griatch** [comment]
The ability to modify database-bound entries from code sounds good in principle ... but - my concern with this is that db-based help entries can also be modified by builders in-game or from the admin. This solution would basically make such changes worthless since they'd always revert to the file-based form, it sounds like? And if you tried the inverse, then the file-based updates would be worthless instead. It sounds like it would make more sense to use Django fixtures for initializing/loading new help entries one time and then allow them to modified in the database. A one-time addition would seem to make more sense for this kind of thing.

----

That said, coincidentally I _did_ add a new `filehelp` functionality to `develop` branch a few weeks ago. My solution does not relate to the database-entries, but makes a new 'type' of help-entry that is file-based (similar to how there are db-based prototypes and module-based ones). It means that a server reload will update all such help entries at once (in the same way as command-help are read from file too). 

See more info in the develop-branch api docs [here](https://www.evennia.com/docs/1.0-dev/api/evennia.help.filehelp.html#module-evennia.help.filehelp) and in new help documentation [here](https://www.evennia.com/docs/1.0-dev/Components/Help-System.html#file-help-entries). While not doing so yet, I'm planning to load these entries (and maybe the command-help entries too) on the default website help/ page so users can read all help entries outside of the game. 
Not being database-bound, they are not modifiable from the admin or from in-game though - only db-based help entries would work for that.

**Reply by davewiththenicehat** [reply]
I started customizing my help system. Remembered you were working on the help system. Wound up switching to the develop branch for file help. What I was doing was too similar to what you were doing for me to bother with my customizations.

If you wind up not having time to make command auto help and help file entries show in the web UI, let me know. I will happily make that a project of mine. It would not hurt me to learn how to make web entries in django.

When database help entries have a matching key in file help entries. The database entry is shown, not the file help entry?

**Comment by Griatch** [comment]
@davewiththenicehat My main updates to the admin and help system in `develop` branch has been completed for now and I'll be moving on to other things. The addition of command/file-help to the front page is something I'll need to get back to later. So if you want to spend time on making them viewable from the website that would be great! Note that the website layout in `evennia/web` has changed a lot compared to master branch though, so make sure you have the latest version. :)

To answer your question, the database entry will override the file-help entry, with the assumption that db-edited files (which can be modified in-game) take precedence over static file-created ones. This way one could have a game with 'pre-made' help files that could then be modified and replaced by active in-game builders.

**Reply by davewiththenicehat** [reply]
That was a big update to the web UI.

Do you foresee issues if I add a locks key to the dictionaries in help_entries.py? To control what is shown on the website.

If I could steal some of your knowledge. Are the `inlines`, where `django.contrib.admin.ModelAdmin` get the data that should show on that page? I haven't looked into `web/website/views/help.py`, guessing that is where the help entries page is managed.

**Comment by Griatch** [comment]
Hah, you are right, the filehelp mechanism needs locks too, I think I just forgot about them :grinning: Yes, the definition should support a `locks` key. I can easily add that (probably tomorrow), you can just assume it will have one for now - it will be using the `view` locktype like the regular db-help entries.

An `inline` is a nested object listing in the admin. 
For this project you should have no reason to modify the admin however - you won't be able to modify the file/command help entries from there anyway, so it serves little purpose to change it. What you need to look at is `evennia/web/website/views/help.py`.

**Reply by Griatch** [reply]
@davewiththenicehat I have now updated `develop` with locks-support for filehelp. Implementing this also lead to some changes of the lock-types (there is a migration to change these for the Db-based HelpEntries). 

- `read` access now determines if a user can read a help entry
- `view` access new determines if the help entry will show up in the help index (this used to be the only thing checked before). If this is not given, the `read` access-type is checked for this as well.
If neither lock is set, the help entry is assumed open for all.

All help-entry types (command, db-based HelpEntry and FileHelpEntry) all use the same locks.

Docs have been updated as well: https://www.evennia.com/docs/1.0-dev/Components/Help-System.html

**Reply by davewiththenicehat** [reply]
@Griatch 
I've done a bit of research.
I think I understand what is needed?

I need to make a model that has all the entries from the three help sources. That model needs to be the attribute `model` in `evennia.web.website.views.HelpMixin.model`.

I need this because I have no way to get a query set for `file help entries` or `command help entries`. I need a complete query set to be return with `evennia.web.website.views.HelpMixin.get_queryset`.

Do you think this would be the best method? Thanks for your time.

**Reply by Griatch** [reply]
No, there is no need to modify the database for this. The `get_queryset` etc is just a convenient way to set up model-searches without needing much code. In this case you are not that helped by this. 

You can look at the `CmdHelp` in `develop` for a mechanism (using the lunr search engine) to query a combined set of db+file+cmd help entries. You may want to use my solution straight off rather than bother with the `get_queryset` stuff, actually.
- https://github.com/evennia/evennia/blob/develop/evennia/commands/default/help.py#L322  - collating help entries from all sources
- https://github.com/evennia/evennia/blob/develop/evennia/commands/default/help.py#L385 - run the search on the candidates using a query
- https://github.com/evennia/evennia/blob/develop/evennia/commands/default/help.py#L746-L766 - abridged implementation in the `sethelp` command (searching to find matches, without the subcategory matching)
